### PR TITLE
feat: [SRM-12476] Update CVNG to handle Error Tracking notifications

### PR DESCRIPTION
**Pull request by ricksjames**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40263)

### CODE REVIEW

File: 980-commons/src/main/java/io/harness/security/ServiceTokenGenerator.java [hunk 1]
```diff
@@ -33,6 +33,7 @@
  */
 @OwnedBy(PL)
 public class ServiceTokenGenerator {
+  public static final AtomicReference<String> ERROR_TRACKING_SERVICE_SECRET = new AtomicReference<>();
```
Comment (cseverino789) [inline 980-commons/src/main/java/io/harness/security/ServiceTokenGenerator.java:36]: Why is required to update the common ServiceTokenGenerator with the ERROR_TRACKING_SERVICE_SECRET? 

Can't CVNG invoke one of the existing methods i.e.

public String getServiceTokenWithDuration(String secretKey, Duration duration)

File: 980-commons/src/main/java/io/harness/security/ErrorTrackingAuthInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  public ErrorTrackingAuthInterceptor(ServiceTokenGenerator tokenGenerator) {
+    this.tokenGenerator = tokenGenerator;
+  }
+
+  @Override
+  public Response intercept(Interceptor.Chain chain) throws IOException {
+    Request request = chain.request();
+    return chain.proceed(request.newBuilder()
+                             .header("Authorization", "ErrorTracking " + tokenGenerator.getErrorTrackingServiceToken())
```
Comment (cseverino789) [inline 980-commons/src/main/java/io/harness/security/ErrorTrackingAuthInterceptor.java:25]: Based on the Comment with the TokenGenerator, anyway to update the the Interceptor with the reference to the ET_SERVICE_SECRET and remove the ET references in the common ServiceTokenGenerator?

File: 980-commons/src/main/java/io/harness/security/ServiceTokenGenerator.java (hunk 1 above)
Comment (ricksjames) [reply 980-commons/src/main/java/io/harness/security/ServiceTokenGenerator.java]: I agree this sounds better, fixed.

File: 980-commons/src/main/java/io/harness/security/ErrorTrackingAuthInterceptor.java (hunk 2 above)
Comment (ricksjames) [reply 980-commons/src/main/java/io/harness/security/ErrorTrackingAuthInterceptor.java]: I agree this sounds better, fixed.

**Review by cseverino789 — DISMISSED** [review]

**Review by praveensugavanam — CHANGES_REQUESTED** [review]
> Reviewed 11 of 25 files at r1, 3 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263)* status: Some of the checks are too old. 13 file(s) to review. 1 change requests. 2. unresolved discussions. 

___
*[`300-cv-nextgen/cv-nextgen-config.yml` line 172 at r2](https://reviewable.io/reviews/harness/harness-core/40263#-NHUu9B91lldjrifCrwX:-NHUu9B91lldjrifCrwY:b-vpgy92) ([raw file](https://github.com/harness/harness-core/blob/badebc6834523180c29269324a75159756260e26/300-cv-nextgen/cv-nextgen-config.yml#L172)):*
> ```YAML
> 
> errorTrackingClientConfig:
>   errorTrackingServiceConfig:
> ```

does this need a special service secret as well ? 
___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/VerificationApplication.java` line 1133 at r2](https://reviewable.io/reviews/harness/harness-core/40263#-NHUuJp_BiCKswfQVutf:-NHUuJp_BiCKswfQVutg:b-ny604u) ([raw file](https://github.com/harness/harness-core/blob/badebc6834523180c29269324a75159756260e26/300-cv-nextgen/src/main/java/io/harness/cvng/VerificationApplication.java#L1133)):*
> ```Java
>     // standardized for service communication.
>     VERIFICATION_SERVICE_SECRET.set(System.getenv(CVNextGenConstants.VERIFICATION_SERVICE_SECRET));
>     ERROR_TRACKING_SERVICE_SECRET.set(System.getenv(CVNextGenConstants.ERROR_TRACKING_SERVICE_SECRET));
> ```

can we get this from the config yaml, similar to manager service secret ? 
___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/client/ErrorTrackingServiceImpl.java` line 15 at r2](https://reviewable.io/reviews/harness/harness-core/40263#-NHUvwlT53Ijy4npk4n5:-NHUvwlT53Ijy4npk4n6:b-118wjz) ([raw file](https://github.com/harness/harness-core/blob/badebc6834523180c29269324a75159756260e26/300-cv-nextgen/src/main/java/io/harness/cvng/client/ErrorTrackingServiceImpl.java#L15)):*
> ```Java
> import java.util.List;
> 
> public class ErrorTrackingServiceImpl implements ErrorTrackingService {
> ```

please add tests

**Review by ricksjames — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263)* status: 44 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`300-cv-nextgen/cv-nextgen-config.yml` line 172 at r2](https://reviewable.io/reviews/harness/harness-core/40263#-NHUu9B91lldjrifCrwX:-NI3rGzM1AhijP6hKqMt:b-rbu3yb) ([raw file](https://github.com/harness/harness-core/blob/badebc6834523180c29269324a75159756260e26/300-cv-nextgen/cv-nextgen-config.yml#L172)):*
<details><summary><i>Previously, praveensugavanam (Praveen Kambam Sugavanam) wrote…</i></summary><blockquote>

does this need a special service secret as well ? 
</blockquote></details>

Fixed. Switched from env variable implementation to configuration implementation. See cv-nextgen-config.yml under srm-service where it was moved to by refactor.
___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/VerificationApplication.java` line 1133 at r2](https://reviewable.io/reviews/harness/harness-core/40263#-NHUuJp_BiCKswfQVutf:-NI3rc-x18hUI20aFEuK:br0q49w) ([raw file](https://github.com/harness/harness-core/blob/badebc6834523180c29269324a75159756260e26/300-cv-nextgen/src/main/java/io/harness/cvng/VerificationApplication.java#L1133)):*
<details><summary><i>Previously, praveensugavanam (Praveen Kambam Sugavanam) wrote…</i></summary><blockquote>

can we get this from the config yaml, similar to manager service secret ? 
</blockquote></details>

Fixed. Added to the configuration yaml
___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/client/ErrorTrackingServiceImpl.java` line 15 at r2](https://reviewable.io/reviews/harness/harness-core/40263#-NHUvwlT53Ijy4npk4n5:-NI3qLkP2pdaWdPWEkS6:bte8xe0) ([raw file](https://github.com/harness/harness-core/blob/badebc6834523180c29269324a75159756260e26/300-cv-nextgen/src/main/java/io/harness/cvng/client/ErrorTrackingServiceImpl.java#L15)):*
<details><summary><i>Previously, praveensugavanam (Praveen Kambam Sugavanam) wrote…</i></summary><blockquote>

please add tests
</blockquote></details>

Done. See ErrorTrackingServiceImplTest.java

**Review by kapil-harness — DISMISSED** [review]
:lgtm:

> Reviewed 1 of 25 files at r1, 1 of 12 files at r2, 43 of 43 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263#-:-NIhpgpsBGJWpxs3CFc2:bnfp4nl)* status: Some of the checks are too old. 1 change requests. 5. unresolved discussions. 

___
*[`948-cv-nextgen-beans/src/main/java/io/harness/cvng/beans/errortracking/Scorecard.java` line 25 at r3](https://reviewable.io/reviews/harness/harness-core/40263#-NIhlB68-0J37z79xa1l:-NIhlB68-0J37z79xa1m:b9fqyyk) ([raw file](https://github.com/harness/harness-core/blob/e7be11d7d7abd51ae39d6818f98d3e863a20d41e/948-cv-nextgen-beans/src/main/java/io/harness/cvng/beans/errortracking/Scorecard.java#L25)):*
> ```Java
> public class Scorecard {
>   String harnessOrgId;
>   String harnessAccountId;
> ```

is prefix harness in harnessAccountId necessary? Also, pls use fully qualified names like accountIdentifier.

**Review by ricksjames — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263)* status: 7 file(s) to review. 1 change requests. 6. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`948-cv-nextgen-beans/src/main/java/io/harness/cvng/beans/errortracking/Scorecard.java` line 25 at r3](https://reviewable.io/reviews/harness/harness-core/40263#-NIhlB68-0J37z79xa1l:-NIi_S0p3En1Sx8FMtt2:b-qssebl) ([raw file](https://github.com/harness/harness-core/blob/e7be11d7d7abd51ae39d6818f98d3e863a20d41e/948-cv-nextgen-beans/src/main/java/io/harness/cvng/beans/errortracking/Scorecard.java#L25)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

is prefix harness in harnessAccountId necessary? Also, pls use fully qualified names like accountIdentifier.
</blockquote></details>

fixed.

**Review by kapil-harness — DISMISSED** [review]
:lgtm:

> Reviewed 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263#-:-NIig3L89_u3X429USO0:bnfp4nl)* status: 1 change requests. 5. unresolved discussions.

**Review by praveensugavanam — APPROVED** [review]
> Reviewed 1 of 25 files at r1, 1 of 12 files at r2, 25 of 43 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263)* status: 2. unresolved discussions.

**Review by ricksjames — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263#-:-NIo4jSC4DTnO6zg64U9:bnfp4nl)* status: 2. unresolved discussions.

**Review by kapil-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263#-:-NIpUO-XASrtmqwBV85C:bnfp4nl)* status: 2. unresolved discussions.

**Review by ricksjames — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263)* status: 2. unresolved discussions. 

___
*[`980-commons/src/main/java/io/harness/security/ErrorTrackingAuthInterceptor.java` line 25 at r1](https://reviewable.io/reviews/harness/harness-core/40263#-NH0yzBrykVWsffHS__r-r1-25:-NIsPgp2B5TpuJOcwzxX:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/1985de633bedc5d60f3b19c63de226b4d766f190/980-commons/src/main/java/io/harness/security/ErrorTrackingAuthInterceptor.java#L25)):*
<details><summary><i>Previously, ricksjames (James Ricks) wrote…</i></summary><blockquote>

I agree this sounds better, fixed.
</blockquote></details>

Done

**Review by cseverino789 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40263)* status: :shipit: complete! ✓

**Pull request merged**